pkg/pulseserver/pulse_server.go

```diff
@@ -0,0 +1,185 @@
+		resourceService: resourceService,
+		sourceClient:    sourceClient,
+	}
+}
+
+// Start initializes and runs the pulse server. It performs the following tasks:
+//
+// 1. Periodically updates the heartbeat for the current maestro instance.
+// 2. Checks the liveness of maestro instances.
+// 3. Subscribes to resource status update messages.
+// 4. Triggers status resync for dead instances.
+//
+// The function runs until the provided context is canceled or an error occurs.
+func (s *pulseServerImpl) Start(ctx context.Context) error {
+	logger := logger.NewOCMLogger(ctx)
+	instance := &api.ServerInstance{
+		Meta: api.Meta{
+			ID:        s.instanceID,
+			UpdatedAt: time.Now(),
+		},
+	}
+	_, err := s.instanceDao.UpSert(ctx, instance)
+	if err != nil {
+		return fmt.Errorf("unable to create maestro instance: %s", err.Error())
+	}
+
+	// start subscribing to resource status update messages.
+	s.startSubscription(ctx)
+
+	pulseTicker := time.NewTicker(time.Duration(s.pulseInterval) * time.Second)
+	checkTicker := time.NewTicker(time.Duration(s.checkInterval) * time.Second)
+	for {
+		select {
+		case <-ctx.Done():
+			pulseTicker.Stop()
+			checkTicker.Stop()
+			return nil
+		case <-pulseTicker.C:
+			logger.V(4).Infof("Updating heartbeat for maestro instance: %s", instance.ID)
+			instance.UpdatedAt = time.Now()
+			_, err = s.instanceDao.UpSert(ctx, instance)
+			if err != nil {
+				// log and ignore the error and continue to tolerate the intermittent issue
+				logger.Error(fmt.Sprintf("Unable to update heartbeat for maestro instance: %s", err.Error()))
+			}
+		case <-checkTicker.C:
+			logger.V(4).Infof("Checking maestro instances liveness and trigger statusresync for dead instances")
+			if err := s.CheckInstances(ctx); err != nil {
+				// log and ignore the error and continue to tolerate the intermittent issue
+				logger.Error(fmt.Sprintf("Unable to check maestro instances: %s", err.Error()))
+			}
+		}
+	}
+}
+
+// CheckInstances checks all maestro instances' liveness and trigger statusresync for dead instances.
+func (s *pulseServerImpl) CheckInstances(ctx context.Context) error {
+	// lock the Instance with a fail-fast advisory lock context.
+	// this allows concurrent processing of many instances by one or more maestro instances exclusively.
+	lockOwnerID, acquired, err := s.lockFactory.NewNonBlockingLock(ctx, "maestro-instances-pulse-check", db.Instances)
+	// Ensure that the transaction related to this lock always end.
+	defer s.lockFactory.Unlock(ctx, lockOwnerID)
+	if err != nil {
+		return fmt.Errorf("error obtaining the event lock: %v", err)
+	}
+	// skip if the lock is not acquired
+	if !acquired {
+		return fmt.Errorf("failed to acquire the lock as another maestro instance is checking instances")
+	}
+	// Instances not pulsing within the last three check intervals are considered as dead.
+	instances, err := s.instanceDao.FindByUpdatedTime(ctx, time.Now().Add(time.Duration(int64(-3*time.Second)*s.checkInterval)))
```

Comment:
hrm why do we still need pulseInterval if we hardcode here.

Reply:
now the instance doesn't pulse with last three times of pulseInterval is inactive.